File: src/graph.py
from dotenv import load_dotenv
load_dotenv()

from langchain_google_genai import ChatGoogleGenerativeAI
from langchain_core.output_parsers import StrOutputParser

chain = ChatGoogleGenerativeAI(model="gemini-1.5-flash") | StrOutputParser()

from typing_extensions import TypedDict, Literal
from langchain_core.runnables import RunnableConfig
from langgraph.graph import StateGraph

# 保持したい情報
class State(TypedDict):
    name: str
    area: str
    feature: str
    names: list[str]

# nodeに使う関数定義（returnした値が更新されます）
def get_name(state: State, config: RunnableConfig):
    return {"name": chain.invoke(f"{state['names']}以外の鳥の名前を1つあげて")}

def get_feature(state: State, config: RunnableConfig):
    return {"feature": chain.invoke(f"{state['name']}の特徴を一言で")}

def get_area(state: State, config: RunnableConfig):
    return {"area": chain.invoke(f"{state['name']}の生息地を一言で")}

def output(state: State, config: RunnableConfig):
    print(state)
    return {"names": state["names"] + [state["name"]]}

def finish(state: State, config: RunnableConfig):
    return {}

# 分岐
def judge(state: State, config: RunnableConfig) -> Literal["name_node", "finish"]:
    return "name_node" if len(state["names"]) < 3 else "finish"

# ワークフロー定義
workflow = StateGraph(State)
# node追加
workflow.add_node("name_node", get_name)
workflow.add_node("feature_node", get_feature)
workflow.add_node("area_node", get_area)
workflow.add_node("output", output)
workflow.add_node("finish", finish)
# edge追加
workflow.add_edge("name_node", "area_node")
workflow.add_edge("name_node", "feature_node")
workflow.add_edge("area_node", "output")
workflow.add_edge("feature_node", "output")
# 始点
workflow.set_entry_point("name_node")
# 終点
workflow.set_finish_point("finish")
# 分岐
workflow.add_conditional_edges("output", judge)

# コンパイル
graph = workflow.compile()
# 実行
print("最終アウトプット:", graph.invoke({"names": []}))
# グラフのmermaid図示
print(graph.get_graph().draw_mermaid())

File: js/graph.js
import { config } from "dotenv";
config();

import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { StringOutputParser } from "@langchain/core/output_parsers";
import { StateGraph } from "@langchain/langgraph";
const chain = new ChatGoogleGenerativeAI({ model: "gemini-1.5-flash" }).pipe(new StringOutputParser())


// nodeに使う関数定義（returnした値が更新されます）
async function getName(state, config) {
    return { name: await chain.invoke(`${state.names}以外の鳥の名前を1つあげて`) };
}

async function getFeature(state, config) {
    return { feature: await chain.invoke(`${state.name}の特徴を一言で`) };
}

async function getArea(state, config) {
    return { area: await chain.invoke(`${state.name}の生息地を一言で`) };
}

async function output(state, config) {
    console.log(state);
    return { names: [...state.names, state.name] };
}

async function finish(state, config) {
    return {};
}

// 分岐
function judge(state, config) {
    return state.names.length < 3 ? "name_node" : "finish";
}

// ワークフロー定義
const workflow = new StateGraph({
    channels: {
        name: { initial: "" },
        area: { initial: "" },
        feature: { initial: "" },
        names: { initial: [] },
    },
});

// node追加
workflow.addNode("name_node", getName);
workflow.addNode("feature_node", getFeature);
workflow.addNode("area_node", getArea);
workflow.addNode("output", output);
workflow.addNode("finish", finish);

// edge追加
workflow.addEdge("name_node", "area_node");
workflow.addEdge("name_node", "feature_node");
workflow.addEdge("area_node", "output");
workflow.addEdge("feature_node", "output");

// 始点
workflow.addEdge("__start__", "name_node");

// 終点
workflow.addEdge("finish", "__end__");

// 分岐
workflow.addConditionalEdges("output", judge);

// コンパイル
const graph = workflow.compile();

const result = await graph.invoke({ names: [] });
console.log("最終アウトプット:", result);
console.log((await graph.getGraphAsync()).drawMermaid());
